Rubberduck.Core/UI/CodeExplorer/Commands/ImportCommand.cs

```diff
@@ -100,43 +116,94 @@ protected override void OnExecute(object parameter)
                 dialog.CheckPathExists = true;
                 dialog.Multiselect = true;
                 dialog.ShowHelp = false;
-                dialog.Title = RubberduckUI.ImportCommand_OpenDialog_Title;
-                dialog.Filter = 
+                dialog.Title = DialogsTitle;
+                dialog.Filter =
                     $"{RubberduckUI.ImportCommand_OpenDialog_Filter_VBFiles} ({_filterExtensions})|{_filterExtensions}|" +
                     $"{RubberduckUI.ImportCommand_OpenDialog_Filter_AllFiles}, (*.*)|*.*";
 
-                if (project == null || dialog.ShowDialog() != DialogResult.OK)
+                if (dialog.ShowDialog() != DialogResult.OK)
                 {
-                    if (usingFreshProjectWrapper)
-                    {
-                        project?.Dispose();
-                    }
-                    return;
+                    return new List<string>();
                 }
 
-                var fileExists = dialog.FileNames.Select(s => s.Split('.').Last());
-                if (fileExists.Any(fileExt => !_importableExtensions.Contains(fileExt)))
+                var fileNames = dialog.FileNames;
+                var fileExtensions = fileNames.Select(Path.GetExtension);
+                if (fileExtensions.Any(fileExt => !_importableExtensions.Contains(fileExt)))
                 {
-                    if (usingFreshProjectWrapper)
-                    {
-                        project.Dispose();
-                    }
-                    return;
+                    NotifyUserAboutAbortDueToUnsupportedFileExtensions(fileNames);
+                    return new List<string>();
                 }
 
-                foreach (var filename in dialog.FileNames)
+                return fileNames;
+            }
+        }
+
+        protected virtual string DialogsTitle => RubberduckUI.ImportCommand_OpenDialog_Title;
+
+        private void NotifyUserAboutAbortDueToUnsupportedFileExtensions(IEnumerable<string> fileNames)
+        {
+            var firstUnsupportedFile = fileNames.First(filename => !_importableExtensions.Contains(Path.GetExtension(filename)));
+            var unsupportedFileName = Path.GetFileName(firstUnsupportedFile);
+            var message = string.Format(RubberduckUI.ImportCommand_UnsupportedFileExtensions, unsupportedFileName);
+            MessageBox.NotifyWarn(message, DialogsTitle);
+        }
+
+        private void ImportFilesWithSuspension(ICollection<string> filesToImport, IVBProject targetProject)
+        {
+            var suspensionResult = _parseManager.OnSuspendParser(this, new[] {ParserState.Ready}, () => ImportFiles(filesToImport, targetProject));
+            if (suspensionResult != SuspensionResult.Completed)
+            {
+                Logger.Warn("File import failed due to suspension failure.");
+            }
+        }
+
+        protected virtual void ImportFiles(ICollection<string> filesToImport, IVBProject targetProject)
```

Comment:
Given that the `Import`/`ImportFromSource` can potentially throw an exception, I think we need to be able to report to the user what was _not_ imported as a result of the exception. Otherwise, the user would be uncertain of the project state should an exception occur?

Reply:
That will complicate things in the module. If somebody wants to enhance the command with this, I am fine with it. However, this will not be part of this PR.